### PR TITLE
feat: using serverless-plugin-warmup to keep some lambdas warm

### PR DESCRIPTION
**Pull request by andreabadesso**
## Motivation
The idea is to keep lambdas that are called from the wallets on load warm to prevent cold starts that take up to 10 seconds per request, causing the wallet load to take a long time.

### Acceptance Criteria
- We should warm only the lambdas that are necessary for the wallet load mechanism:
  - `loadWalletAsync`
  - `loadWalletApi`
  - `getWalletStatusApi`
  - `getAddressesApi`
  - `getNewAddresses`
  - `getBalanceApi`
  - `getTokensApi`
  - `getVersionData`
  - `getTxHistoryApi`

### Security Checklist
- [X] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: serverless.yml [hunk 1]
```diff
@@ -127,10 +129,12 @@ functions:
   onNewTxEvent:
     handler: src/txProcessor.onNewTxEvent
   loadWalletAsync:
+    warmup: true
     handler: src/api/wallet.loadWallet
   loadWalletApi:
     role: arn:aws:iam::769498303037:role/WalletServiceLoadWalletLambda
     handler: src/api/wallet.load
+    warmup: true
```
Comment (luislhl) [inline serverless.yml:141]: Is it possible to detect from inside the function that it's just being warmed? This way we could avoiding performing any real work in the code.

File: serverless.yml [hunk 2]
```diff
@@ -213,13 +222,15 @@ functions:
...
     handler: src/api/version.get
+    warmup: true
     events:
       - http:
           path: version
           method: get
           cors: true
   getTxHistoryApi:
     handler: src/api/txhistory.get
+    warmup: true
```
Comment (luislhl) [inline serverless.yml:237]: On which basis did you select what functions need warmup?

File: serverless.yml (hunk 1 above)
Comment (andreabadesso) [reply serverless.yml:141]: It is, that is what I'm working right now

I want to use `middy` to detect when it's a warmup call (`event.source === 'serverless-plugin-warmup'`) and return immediately, saving some processing time

File: serverless.yml (hunk 2 above)
Comment (andreabadesso) [reply serverless.yml:237]: This is still a WIP, but I want to warmup all functions related to the app startup, I will document this on the PR before sending it to review

**Comment by luislhl** [comment]
> The idea is to keep lambdas that are called from the wallets on load warm to prevent cold starts that take up to 10 seconds per request, causing the wallet load to take a long time.

This would probably solve for almost all cases right now, because we don't have so many users yet.

But at some point in the future when we have more users, if we have more than 1 person at the same time making requests to load the wallet, the lambda warm-up would only solve the cold-starts for 1 of them, right?

>  take up to 10 seconds per request

Did you measure this or is this based on some documentation?

**Comment by andreabadesso** [comment]
@luislhl 

> But at some point in the future when we have more users, if we have more than 1 person at the same time making requests to load the wallet, the lambda warm-up would only solve the cold-starts for 1 of them, right?

The serverless-plugin-warmup has a `concurrency` option that calls the lambdas in parallel, as per documentation:

```
concurrency: The number of times that each of your lambda functions will be called in parallel. This can be used in a best-effort attempt to force AWS to spin up more parallel containers for your lambda. (defaults to 1)
```

When we start the rollout we can monitor the service to check if the lambdas are being cold started and increase this value. What do you think?

> Did you measure this or is this based on some documentation?

Just checked again the cold start times and it is averaging at 1-2s. But it is based on my observation only, not on any documentation from aws

**Comment by luislhl** [comment]
> When we start the rollout we can monitor the service to check if the lambdas are being cold started and increase this value. What do you think?

How could we have an alert that warn us if they are being cold started?

One possible thing to monitor could be the start time of the wallets themselves. Monitoring this on the clients could generate privacy concerns, but maybe we could do this in the backend somehow. This way we could have alerts for when too many wallets are taking too long.

Another thing that is even easier is to monitor the `Concurrent executions` metric of the Lambdas, so we detect when we start having too many concurrent executions, which means we would benefit from increasing the `concurrency` option in the plugin.

What do you think?

Maybe these items could be added to the issue we have about monitoring the wallet-service

**Review by luislhl — APPROVED** [review]

**Review by r4mmer — APPROVED** [review]

**Pull request merged**